File: rd-net/RdFramework/Impl/RdProperty.cs
using System;
using System.ComponentModel;
using System.Threading;
using JetBrains.Collections.Viewable;
using JetBrains.Core;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;
using JetBrains.Annotations;
using JetBrains.Util.Internal;

namespace JetBrains.Rd.Impl
{
  public abstract class RdPropertyBase : RdReactiveBase {}
  public class RdProperty<T> : RdPropertyBase, IViewableProperty<T>
  {
    #region Constructor

    public override event PropertyChangedEventHandler? PropertyChanged;

    public RdProperty(CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
    {
      ReadValueDelegate = readValue;
      WriteValueDelegate = writeValue;

      Advise(Lifetime.Eternal, _ =>
      {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("Value"));
      });
    }

    [PublicAPI]
    public RdProperty(CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue, T defaultValue) : this(readValue, writeValue)
    {
      Value = defaultValue;
    }

    #endregion


    #region Serializers

    public CtxReadDelegate<T> ReadValueDelegate { get; private set; }
    public CtxWriteDelegate<T> WriteValueDelegate { get; private set; }

    public static RdProperty<T> Read(SerializationCtx ctx, UnsafeReader reader)
    {
      return Read(ctx, reader, Polymorphic<T>.Read, Polymorphic<T>.Write);
    }

    public static RdProperty<T> Read(SerializationCtx ctx, UnsafeReader reader, CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
    {
      var id = reader.ReadRdId();
      var res =  new RdProperty<T>(readValue, writeValue).WithId(id);
      if (reader.ReadBool())
      {
        res.myProperty.Value = readValue(ctx, reader);
      }
      return res;
    }

    public static void Write(SerializationCtx ctx, UnsafeWriter writer, RdProperty<T> value)
    {
      Assertion.Assert(!value.RdId.IsNil);
      writer.Write(value.RdId);
      if (value.HasValue())
      {
        writer.WriteBoolean(true);
        value.WriteValueDelegate(ctx, writer, value.Value);
      }
      else
      {
        writer.WriteBoolean(false);
      }
    }

    #endregion
    #region Mastering

    public bool IsMaster = false;
    private int myMasterVersion;

    #endregion
    #region Init

    public bool OptimizeNested = false;
    private LifetimeDefinition? myBindDefinition;

    public override void Identify(IIdentities identities, RdId id)
    {
      base.Identify(identities, id);
      if (!OptimizeNested)
      {
        Maybe.ValueOrDefault.IdentifyPolymorphic(identities, identities.Next(id));
      }
    }

    protected override void PreInit(Lifetime lifetime, IProtocol proto)
    {
      base.PreInit(lifetime, proto);

      if (!OptimizeNested)
      {
        var maybe = Maybe;
        if (maybe.HasValue)
        {
          var definition = TryPreBindValue(lifetime, maybe.Value, false);
          using var cookie = lifetime.UsingExecuteIfAlive();
          if (cookie.Succeed)
          {
            var prevDefinition = Interlocked.Exchange(ref myBindDefinition , definition);
            Assertion.Assert(prevDefinition?.Lifetime.IsNotAlive ?? true);

          }
          else return;
        }
      }

      proto.Wire.Advise(lifetime, this);
    }

    protected override void Init(Lifetime lifetime, IProtocol proto, SerializationCtx ctx)
    {
      base.Init(lifetime, proto, ctx);

      var maybe = Maybe;
      var hasInitValue = maybe.HasValue;
      if (hasInitValue && !OptimizeNested)
        maybe.Value.BindPolymorphic();

      Advise(lifetime, v =>
      {
        var shouldIdentify = !hasInitValue;
        hasInitValue = false;

        if (!IsLocalChange)
          return;

        if (!OptimizeNested && shouldIdentify)
        {
          // We need to terminate the current lifetime to unbind the existing value before assigning a new value, especially in cases where we are reassigning it.
          Memory.VolatileRead(ref myBindDefinition)?.Terminate();

          v.IdentifyPolymorphic(proto.Identities, proto.Identities.Next(RdId));

          var prevDefinition = Interlocked.Exchange(ref myBindDefinition, TryPreBindValue(lifetime, v, false));
          prevDefinition?.Terminate();
        }

        if (IsMaster) myMasterVersion++;

        proto.Wire.Send(RdId, SendContext.Of(ctx, v, this), static (sendContext, writer) =>
        {
          var sContext = sendContext.SzrCtx;
          var evt = sendContext.Event;
          var me = sendContext.This;
          writer.WriteInt32(me.myMasterVersion);
          me.WriteValueDelegate(sContext, writer, evt);
          SendTrace?.Log($"{me} :: ver = {me.myMasterVersion}, value = {me.Value.PrintToString()}");
        });

        if (!OptimizeNested && shouldIdentify)
          v.BindPolymorphic();
      });
    }

    protected override void Unbind()
    {
      base.Unbind();
      myBindDefinition = null;
    }

    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader reader, IRdWireableDispatchHelper dispatchHelper)
    {
      var version = reader.ReadInt();
      var value = ReadValueDelegate(ctx, reader);

      var lifetime = dispatchHelper.Lifetime;
      var definition = TryPreBindValue(lifetime, value, true);

      ReceiveTrace?.Log($"OnWireReceived:: {GetMessage(version, value)}");

      dispatchHelper.Dispatch(() =>
      {
        var rejected = IsMaster && version < myMasterVersion;

        ReceiveTrace?.Log($"Dispatch:: {GetMessage(version, value)}{(rejected ? " REJECTED" : "")}");

        if (rejected)
        {
          definition?.Terminate();
          return;
        }

        myMasterVersion = version;

        using (UsingDebugInfo())
        {
          Interlocked.Exchange(ref myBindDefinition, definition)?.Terminate();
          myProperty.Value = value;
        }
      });
    }

    private string GetMessage(int version, T value)
    {
      return $"{this} :: oldver = {myMasterVersion}, newver = {version}, value = {value.PrintToString()}";
    }

    private LifetimeDefinition? TryPreBindValue(Lifetime lifetime, T value, bool bindAlso)
    {
      if (OptimizeNested || !value.IsBindable())
        return null;

      var definition = new LifetimeDefinition { Id = value };
      try
      {
        value.PreBindPolymorphic(definition.Lifetime, this, "$");
        if (bindAlso)
          value.BindPolymorphic();

        lifetime.Definition.Attach(definition, true);
        return definition;
      }
      catch
      {
        definition.Terminate();
        throw;
      }
    }

    #endregion


    #region Api

    private readonly IViewableProperty<T> myProperty = new ViewableProperty<T>();

    public ISource<T> Change => myProperty.Change;

    public Maybe<T> Maybe => myProperty.Maybe;

    public T Value
    {
      get => myProperty.Value;
      set
      {
        AssertNullability(value);
        using (UsingLocalChange())
        {
          myProperty.Value = value;
        }
      }
    }

    public void Advise(Lifetime lifetime, Action<T> handler)
    {
      if (IsBound) AssertThreading();

      using (UsingDebugInfo())
        myProperty.Advise(lifetime, handler);
    }

    public override RdBindableBase? FindByRName(RName rName)
    {
      var rootName = rName.GetNonEmptyRoot();
      var localName = rootName.LocalName.ToString();
      if (localName != "$")
        return null;

      var maybe = myProperty.Maybe;
      if (!maybe.HasValue)
        return null;

      if (!(maybe.Value is RdBindableBase value))
        return null;

      if (rootName == rName)
        return value;

      return value.FindByRName(rName.DropNonEmptyRoot());
    }

    #endregion


    protected override string ShortName => "property";

    public override void Print(PrettyPrinter printer)
    {
      base.Print(printer);

      if (!printer.PrintContent)
        return;

      printer.Print("(ver=" + myMasterVersion + ") [");
      if (Maybe.HasValue)
      {
        using (printer.IndentCookie())
        {
          Value.PrintEx(printer);
        }
      }
      else
      {
        printer.Print(" <not initialized> ");
      }

      printer.Print("]");
    }


  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/RdProperty.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.LifetimeDefinition
import com.jetbrains.rd.util.lifetime.isNotAlive
import com.jetbrains.rd.util.reactive.*
import com.jetbrains.rd.util.string.*
import com.jetbrains.rd.util.trace
import java.util.concurrent.atomic.AtomicReference

abstract class RdPropertyBase<T>(val valueSerializer: ISerializer<T>) : RdReactiveBase(), IMutablePropertyBase<T> {
    companion object {
//        @JvmStatic
        fun<T : Any> write0(ctx: SerializationCtx, buffer: AbstractBuffer, prop: RdPropertyBase<T>, value: T?) {
            prop.rdid.notNull().write(buffer)
            if (value != null) {
                buffer.writeBool(true)
                prop.valueSerializer.write(ctx, buffer, value)
            } else {
                buffer.writeBool(false)
            }
        }
    }

    //mastering
    protected var masterVersion = 0

    var defaultValueChanged : Boolean = false
        internal set(value) {
            field = value
        }
    //init
    var optimizeNested: Boolean = false
    private val bindDefinition: AtomicReference<LifetimeDefinition?> = AtomicReference(null)

    protected abstract val property: IMutablePropertyBase<T>

    override val change : ISource<T> get() = property.change

    protected abstract val valueOrNull: T?

    override fun preInit(lifetime: Lifetime, proto: IProtocol) {
        super.preInit(lifetime, proto)

        if (!optimizeNested) {
            val value = valueOrNull
            if (value != null) {
                val definition = tryPreBindValue(lifetime, value, false)
                lifetime.executeIfAlive {
                    val prevDefinition = bindDefinition.getAndSet(definition)
                    assert(prevDefinition?.isNotAlive ?: true)
                }

            }
        }

        proto.wire.advise(lifetime, this)
    }

    override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
        super.init(lifetime, proto, ctx)

        val value = valueOrNull
        var hasInitValue = value != null
        if (hasInitValue && !optimizeNested)
            value.bindPolymorphic()

        advise(lifetime) { v ->
            val shouldIdentify = !hasInitValue
            hasInitValue = false

            if (!isLocalChange)
                return@advise

            if (!optimizeNested && shouldIdentify) {
                // We need to terminate the current lifetime to unbind the existing value before assigning a new value, especially in cases where we are reassigning it.
                bindDefinition.get()?.terminate()

                v.identifyPolymorphic(proto.identity, proto.identity.next(rdid))

                val prevDefinition = bindDefinition.getAndSet(tryPreBindValue(lifetime, v, false))
                prevDefinition?.terminate()
            }

            if (master) masterVersion++

            proto.wire.send(rdid) { buffer ->
                buffer.writeInt(masterVersion)
                valueSerializer.write(ctx, buffer, v)
                logSend.trace { "property `$location` ($rdid):: ver = $masterVersion, value = ${v.printToString()}" }
            }

            if (!optimizeNested && shouldIdentify)
                v.bindPolymorphic()
        }
    }

    override fun unbind() {
        super.unbind()
        bindDefinition.getAndSet(null)
    }

    override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
        val version = buffer.readInt()
        val v = valueSerializer.read(ctx, buffer)

        val definition = tryPreBindValue(dispatchHelper.lifetime, v, true)

        logReceived.trace { "onWireReceived:: ${getMessage(version, v)}" }

        dispatchHelper.dispatch {
            val rejected = master && version < masterVersion
            logReceived.trace { "dispatched:: ${getMessage(version, v)}${rejected.condstr { " >> REJECTED" }}" }

            if (rejected) {
                definition?.terminate()
                return@dispatch
            }

            masterVersion = version

            val prevDefinition = bindDefinition.getAndSet(definition)
            prevDefinition?.terminate()
            property.set(v)
        }
    }

    private fun getMessage(version: Int, v: T) =
        "property `$location` ($rdid):: oldver = $masterVersion, newver = $version, value = ${v.printToString()}"

    override fun advise(lifetime: Lifetime, handler: (T) -> Unit) {
        if (isBound) assertThreading()
        property.advise(lifetime, handler)
    }

    private fun tryPreBindValue(lifetime: Lifetime, value: T, bindAlso: Boolean): LifetimeDefinition? {
        if (optimizeNested || value == null)
            return null

        val definition = LifetimeDefinition().apply { id = value }
        try {
            value.preBindPolymorphic(definition.lifetime, this, "$")
            if (bindAlso)
                value.bindPolymorphic()

            lifetime.attach(definition, true)
            return definition
        } catch (e: Throwable) {
            definition.terminate()
            throw e
        }
    }
}



@Suppress("UNCHECKED_CAST")
class RdOptionalProperty<T : Any>(valueSerializer: ISerializer<T> = Polymorphic())
    : RdPropertyBase<T>(valueSerializer), IOptProperty<T> {

    override fun deepClone(): RdOptionalProperty<T> = RdOptionalProperty(valueSerializer).also { if (hasValue) it.set(valueOrThrow.deepClonePolymorphic()) }

    //constructor
    constructor(defaultValue: T, valueSerializer: ISerializer<T> = Polymorphic()) : this(valueSerializer) {
        set(defaultValue)
    }

    //serializers
    companion object : ISerializer<RdOptionalProperty<*>> {

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdOptionalProperty<*> = read(ctx, buffer, Polymorphic())
        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdOptionalProperty<*>) {
            val value1 = value.valueOrNull
            write0(ctx, buffer, value as RdPropertyBase<Any>, value1)
        }

        fun <T : Any> read(ctx: SerializationCtx, buffer: AbstractBuffer, valueSerializer: ISerializer<T>): RdOptionalProperty<T> {
            val id = RdId.read(buffer)
            val res = RdOptionalProperty(valueSerializer).withId(id)
            if (buffer.readBool()) {
                res.property.set(valueSerializer.read(ctx, buffer))
            }
            return res
        }

    }

    fun slave() : RdOptionalProperty<T> {
        master = false
        return this
    }


    //api
    override val property = OptProperty<T>()
    override val valueOrNull: T?
        get() = property.valueOrNull

    override fun set(newValue: T) {
        localChange {
            defaultValueChanged = true
            property.set(newValue)
        }
    }

    override fun identify(identities: IIdentities, id: RdId) {
        super.identify(identities, id)
        if (!optimizeNested)
            valueOrNull?.identifyPolymorphic(identities, identities.next(id))
    }

    override fun advise(lifetime: Lifetime, handler: (T) -> Unit) {
        super<RdPropertyBase>.advise(lifetime, handler)
    }

    override fun findByRName(rName: RName): RdBindableBase? {
        if (rName == RName.Empty) return this
        val rootName = rName.getNonEmptyRoot()
        val localName = rootName.localName
        if (localName != "$")
            return null

        val value = property.valueOrNull as? RdBindableBase
            ?: return null

        if (rootName == rName)
            return value

        return value.findByRName(rName.dropNonEmptyRoot())
    }

    //pretty printing
    override fun print(printer: PrettyPrinter) {
        super.print(printer)
        printer.print("(ver=$masterVersion) [")
        valueOrNull.let {
            when (it) {
                null -> printer.print(" <not initialized> ")
                else -> printer.indent { it.print(printer) }
            }
            Unit
        }
        printer.print("]")
    }
}

@Suppress("UNCHECKED_CAST")
class RdProperty<T>(defaultValue: T, valueSerializer: ISerializer<T> = Polymorphic())
    : RdPropertyBase<T>(valueSerializer), IProperty<T> {

    override fun deepClone(): RdProperty<T> = RdProperty(value.deepClonePolymorphic(), valueSerializer)

    //serializers
    companion object : ISerializer<RdProperty<*>> {

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdProperty<*> = read(ctx, buffer, Polymorphic<Any?>())
        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdProperty<*>) =
            write0(ctx, buffer, value as RdPropertyBase<Any>, value.value)

        fun <T> read(ctx: SerializationCtx, buffer: AbstractBuffer, valueSerializer: ISerializer<T>): RdProperty<T> {
            val id = RdId.read(buffer)
            val value = if (buffer.readBool()) valueSerializer.read(ctx, buffer) else null
            return RdProperty(value as T, valueSerializer).withId(id)
        }
    }

    fun slave() : RdProperty<T> {
        master = false
        return this
    }

    override val valueOrNull: T?
        get() = value

    //api
    override val property = Property(defaultValue)

    override var value: T
        get() = property.value
        set(value) = localChange {
            defaultValueChanged = true
            property.value = value
        }

    override fun set(newValue: T) {
        value = newValue
    }

    override fun identify(identities: IIdentities, id: RdId) {
        super.identify(identities, id)
        if (!optimizeNested)
            value?.identifyPolymorphic(identities, identities.next(id))
    }

    override fun advise(lifetime: Lifetime, handler: (T) -> Unit) {
        super<RdPropertyBase>.advise(lifetime, handler)
    }


    override fun findByRName(rName: RName): RdBindableBase? {
        if (rName == RName.Empty) return this
        val rootName = rName.getNonEmptyRoot()
        val localName = rootName.localName
        if (localName != "$")
            return null

        val value = property.value as? RdBindableBase
            ?: return null

        if (rootName == rName)
            return value

        return value.findByRName(rName.dropNonEmptyRoot())
    }

    //pretty printing
    override fun print(printer: PrettyPrinter) {
        super.print(printer)
        printer.print("(ver=$masterVersion) [")
        value.let {
            when (it) {
                null -> printer.print(" <null> ")
                else -> printer.indent { it.print(printer) }
            }
            Unit
        }
        printer.print("]")
    }
}
